src/storage/src/storage/upload_object.rs

```diff
@@ -525,30 +527,76 @@ impl<T> UploadObject<T> {
+    ///     .with_crc32c(crc32c(b"hello world"))
+    ///     .send()
+    ///     .await?;
+    /// println!("response details={response:?}");
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// In some applications, the payload's CRC32C checksum is already known.
+    /// For example, the application may be downloading the data from another
+    /// blob storage system.
+    ///
+    /// In such cases, it is safer to pass the known CRC32C of the payload to
+    /// [Cloud Storage], and more efficient to skip the computation in the
+    /// client library.
+    ///
+    /// Note that once you provide a CRC32C value to this builder you cannot
+    /// enable additional checksums.
+    pub fn with_crc32c<V>(mut self, v: V) -> UploadObject<T, Precomputed>
     where
         V: Into<u32>,
     {
-        let mut checksum = self.mut_resource().checksums.take().unwrap_or_default();
+        let checksum = self.mut_resource().checksums.get_or_insert_default();
         checksum.crc32c = Some(v.into());
-        self.mut_resource().checksums = Some(checksum);
-        self
+        self.switch_checksum(|_| Precomputed)
     }
 
-    // TODO(#2050) - this should be automatically computed?
-    #[allow(dead_code)]
-    pub(crate) fn with_md5_hash<I, V>(mut self, i: I) -> Self
+    /// Provide a precomputed value for the MD5 hash.
+    ///
+    /// # Example
+    /// ```
+    /// # use google_cloud_storage::client::Storage;
+    /// # async fn sample(client: &Storage) -> anyhow::Result<()> {
+    /// use md5::compute;
+    /// let hash = md5::compute(b"hello world");
+    /// let response = client
+    ///     .upload_object("projects/_/buckets/my-bucket", "my-object", "hello world")
+    ///     .with_md5_hash(bytes::Bytes::from_owner(hash.0))
+    ///     .send()
+    ///     .await?;
+    /// println!("response details={response:?}");
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// In some applications, the payload's MD5 hash is already known. For
+    /// example, the application may be downloading the data from another blob
+    /// storage system.
+    ///
+    /// In such cases, it is safer to pass the known MD5 of the payload to
+    /// [Cloud Storage], and more efficient to skip the computation in the
+    /// client library.
+    ///
+    /// Note that once you provide a CRC32C value to this builder you cannot
+    /// enable additional checksums.
+    pub fn with_md5_hash<I, V>(mut self, i: I) -> UploadObject<T, Precomputed>
     where
         I: IntoIterator<Item = V>,
         V: Into<u8>,
     {
-        let mut checksum = self.mut_resource().checksums.take().unwrap_or_default();
+        let checksum = self.mut_resource().checksums.get_or_insert_default();
         checksum.md5_hash = i.into_iter().map(|v| v.into()).collect();
-        // TODO(#2050) - should we return an error (or panic?) if the size is wrong?
-        self.mut_resource().checksums = Some(checksum);
-        self
+        self.switch_checksum(|_| Precomputed)
```

Comment:
maybe I am confused, but won't this also set `crc32c` to be precomputed?

how do I ask the client library to use the precomputed `md5` hash I give it and compute its own `crc32c` ?

Reply:
I think we could relax this in the future by introducing `PrecomputedCrc32c` and `PrecomputedMd5`, where each of those would still allow you to still compute the other value.